File: tscale/archive/old_implementations/ascii_scope.py
#!/usr/bin/env python3
# ascii_scope.py — 4-channel ASCII scope for FIFO/stdin
# Input:  t ch1 ch2 ch3 ch4   (>=2 cols; whitespace-sep)
# Keys: q quit | space run/stop | </> timebase | t trig | g trig ch | +/- trig lvl
#       r/f/e edge | o envelope/points | m auto-span | 1..4 toggle | A/S/D/F gain± | z/x/c/v off±

import sys, os, time, curses, select, signal, statistics
from collections import deque

# ---------- config ----------
PATH             = sys.argv[1] if len(sys.argv) > 1 else "/tmp/scope.fifo"
MAX_CH           = 4
REFRESH_HZ       = 30
HIST_SEC         = 2.0
DEFAULT_SPAN     = 0.250
STACK_FRAC       = 0.45
TRIG_POS_FRAC    = 0.25
DIGITAL_LAST     = False

# ---------- state ----------
class Chan:
    def __init__(self, color_idx):
        self.visible = True
        self.gain = 1.0
        self.offset = 0.0
        self.color = color_idx

chans = [Chan((i % 6) + 1) for i in range(MAX_CH)]
buf = deque(maxlen=300000)   # (t, [v...])
last_t = None

run = True
time_span_s = DEFAULT_SPAN
auto_span = True              # m toggles (adapts span to fill columns)
trig_enabled = False          # default off to avoid locked column
trig_ch = 1
trig_lvl = 0.0
trig_edge = +1
trig_hyst = 1e-9
use_envelope = True

fd = None
is_stdin = PATH in ("-", "/dev/stdin")

# ---------- io ----------
def open_input():
    global fd
    if is_stdin:
        fd = sys.stdin.fileno()
        os.set_blocking(fd, False)
        return
    if not os.path.exists(PATH):
        raise SystemExit(f"FIFO not found: {PATH}\nCreate with: mkfifo {PATH}")
    fd = os.open(PATH, os.O_RDONLY | os.O_NONBLOCK)

def parse_line(line):
    try:
        parts = line.strip().split()
        if len(parts) < 2: return None
        t = float(parts[0])
        vals = [float(p) for p in parts[1:1+MAX_CH]]
        return (t, vals)
    except Exception:
        return None

def read_into_buffer():
    global last_t
    if fd is None: return 0
    r, _, _ = select.select([fd], [], [], 0)
    n = 0
    if fd in r:
        try:
            data = os.read(fd, 65536).decode('utf-8', 'ignore') if not is_stdin else sys.stdin.read()
        except (BlockingIOError, InterruptedError):
            data = ""
        if not data:
            if not is_stdin:
                try: os.close(fd)
                except Exception: pass
                time.sleep(0.01); open_input()
            return 0
        for line in data.splitlines():
            rec = parse_line(line)
            if not rec: continue
            t, vs = rec
            if last_t is not None and t < last_t:
                t = last_t + 1e-12
            last_t = t
            buf.append((t, vs)); n += 1
    if n and last_t is not None:
        cutoff = last_t - max(HIST_SEC, 5*DEFAULT_SPAN)
        while buf and buf[0][0] < cutoff: buf.popleft()
    return n

# ---------- window ----------
def estimate_dt():
    """Robust dt estimate from last ~200 samples (guards against clumping)."""
    if len(buf) < 5: return None
    ts = [buf[i][0] for i in range(max(0, len(buf)-200), len(buf))]
    dts = [b-a for a,b in zip(ts, ts[1:]) if b>a]
    if not dts: return None
    return statistics.median(dts)

def compute_window(now_t, w_cols):
    global time_span_s
    if not buf: return (now_t - time_span_s, now_t)

    # optional trigger
    if trig_enabled and (1 <= trig_ch <= MAX_CH):
        ch = trig_ch - 1
        lo = trig_lvl - trig_hyst; hi = trig_lvl + trig_hyst
        lower = buf[-1][0] - 5.0*time_span_s
        for i in range(len(buf)-1, 0, -1):
            t2, v2s = buf[i]; t1, v1s = buf[i-1]
            if t2 < lower: break
            if ch >= len(v1s) or ch >= len(v2s): continue
            v1 = chans[ch].gain*v1s[ch] + chans[ch].offset
            v2 = chans[ch].gain*v2s[ch] + chans[ch].offset
            if (trig_edge > 0 and (v1 <= lo and v2 >= hi)) or \
               (trig_edge < 0 and (v1 >= hi and v2 <= lo)) or \
               (trig_edge == 0 and ((v1 <= lo and v2 >= hi) or (v1 >= hi and v2 <= lo))):
                left = t2 - TRIG_POS_FRAC*time_span_s
                return (left, left + time_span_s)

    # auto-span: fill at least ~70% of columns with data buckets
    if auto_span and len(buf) > 2:
        dt = estimate_dt()
        if dt:
            # target ~0.8 * visible columns, with safety clamp
            target_cols = max(10, int(0.8 * max(2, w_cols)))
            span = max(5*dt, target_cols * dt)
            time_span_s = max(0.005, min(60.0, span))

    right = buf[-1][0]
    return (right - time_span_s, right)

# ---------- mapping ----------
def map_to_row(val, rows, ci):
    sep = rows / (MAX_CH + 1)
    mid = int((ci+1)*sep)
    scale = sep * STACK_FRAC
    y = int(mid - val*scale)
    if y < 0: y = 0
    if y >= rows: y = rows-1
    return y

# ---------- renderers ----------
def render_envelopes(stdscr, left_t, right_t, h, w):
    cols = max(2, w-2)
    rows = h-1
    span = max(1e-12, right_t - left_t)
    for ci, ch in enumerate(chans):
        if not ch.visible: continue
        attr = curses.color_pair(ch.color)
        ymin = [None]*cols; ymax = [None]*cols
        # bin to nearest column center to avoid left-edge pileup
        for t, vs in buf:
            if t < left_t or t > right_t: continue
            if ci >= len(vs): continue
            xf = (cols-1) * (t - left_t) / span
            x = int(round(xf))                      # round() fixes single-column artifacts
            if x < 0 or x >= cols: continue
            yv = ch.gain*vs[ci] + ch.offset
            if DIGITAL_LAST and ci == (MAX_CH-1):
                yv = 0.8 if yv >= 0.5 else -0.8
            y = map_to_row(yv, rows, ci)
            if ymin[x] is None or y < ymin[x]: ymin[x] = y
            if ymax[x] is None or y > ymax[x]: ymax[x] = y
        # draw
        for x in range(cols):
            y0 = ymin[x]
            if y0 is None: continue
            y1 = ymax[x] if ymax[x] is not None else y0
            X = 1 + x
            if y0 == y1:
                stdscr.addch(y0, X, '*', attr)
            else:
                step = 1 if y1 >= y0 else -1
                for ry in range(y0, y1+step, step):
                    stdscr.addch(ry, X, '|', attr)

def render_points(stdscr, left_t, right_t, h, w):
    rows = h-1
    span = max(1e-12, right_t - left_t)
    for ci, ch in enumerate(chans):
        if not ch.visible: continue
        attr = curses.color_pair(ch.color)
        last_x = None; last_y = None
        for t, vs in buf:
            if t < left_t or t > right_t: continue
            if ci >= len(vs): continue
            xf = 1 + (w-2) * (t - left_t) / span
            x = int(round(xf))                    # use round()
            if x < 1 or x >= w-1: continue
            yv = ch.gain*vs[ci] + ch.offset
            if DIGITAL_LAST and ci == (MAX_CH-1):
                yv = 0.8 if yv >= 0.5 else -0.8
            y = map_to_row(yv, rows, ci)
            stdscr.addch(y, x, '*', attr)
            if last_x is not None and x > last_x:
                dx = x - last_x; dy = y - last_y
                for i in range(1, dx):
                    xi = last_x + i
                    yi = last_y + (dy*i)//dx
                    if 1 <= xi < w-1 and 0 <= yi < rows:
                        stdscr.addch(yi, xi, '.', attr)
            last_x, last_y = x, y

# ---------- ui ----------
def header_line():
    run_s = "ON" if run else "OFF"
    edge_s = "+" if trig_edge>0 else "-" if trig_edge<0 else "+/-"
    return "  ".join([
        f"[q] quit  [space] run={run_s}",
        f"[</>] timebase {time_span_s:.3f}s  [m] auto={'ON' if auto_span else 'OFF'}",
        f"[t] trigger {'ON' if trig_enabled else 'OFF'} ch={trig_ch} lvl={trig_lvl:+.3g} edge={edge_s}",
        "[1..4] toggle  [A/S/D/F] gain±  [z/x/c/v] off±  [o] env/pts"
    ])

def draw_scope(stdscr):
    global run, time_span_s, auto_span, trig_enabled, trig_ch, trig_lvl, trig_edge, use_envelope
    curses.curs_set(0)
    stdscr.nodelay(True)
    stdscr.timeout(int(1000/REFRESH_HZ))
    if curses.has_colors():
        curses.start_color()
        for i in range(1, 8): curses.init_pair(i, i, 0)

    open_input()
    while True:
        read_into_buffer()
        h, w = stdscr.getmaxyx()
        stdscr.erase()

        # header
        stdscr.addnstr(0, 0, header_line(), w-1, curses.A_REVERSE)

        # axes
        for y in range(1, h): stdscr.addch(y, 0, '|')
        stdscr.addnstr(h-1, 0, '+' + '-'*(w-2) + '>', w-1)

        # window
        now_t = buf[-1][0] if buf else time.time()
        left_t, right_t = compute_window(now_t, w-2)
        if right_t <= left_t: right_t = left_t + 1e-3

        # render
        if use_envelope:
            render_envelopes(stdscr, left_t, right_t, h, w)
        else:
            render_points(stdscr, left_t, right_t, h, w)

        # footer
        meta = [f"ch{i}:{'on' if ch.visible else 'off'} g={ch.gain:.3g} off={ch.offset:+.3g}"
                for i, ch in enumerate(chans, 1)]
        stdscr.addnstr(h-2, 2, "  ".join(meta), w-4)

        stdscr.refresh()

        # keys
        c = stdscr.getch()
        if c == -1: continue
        if c in (ord('q'), 27): return
        elif c == ord(' '): run = not run
        elif c in (curses.KEY_RIGHT, ord('>'), ord('.')): time_span_s = min(60.0, time_span_s*1.25)
        elif c in (curses.KEY_LEFT,  ord('<'), ord(',')): time_span_s = max(0.005, time_span_s/1.25)
        elif c == ord('m'): auto_span = not auto_span
        elif c == ord('t'): trig_enabled = not trig_enabled
        elif c == ord('g'): trig_ch = 1 + (trig_ch % MAX_CH)
        elif c in (ord('+'), ord('=')): trig_lvl += 0.05
        elif c == ord('-'): trig_lvl -= 0.05
        elif c == ord('r'): trig_edge = +1
        elif c == ord('f'): trig_edge = -1
        elif c == ord('e'): trig_edge = 0
        elif c == ord('o'): use_envelope = not use_envelope
        elif c in (ord('1'), ord('2'), ord('3'), ord('4')):
            idx = c - ord('1'); chans[idx].visible = not chans[idx].visible
        elif c in (ord('a'), ord('A')): chans[0].gain *= (1.1 if c==ord('A') else 1/1.1)
        elif c in (ord('s'), ord('S')): chans[1].gain *= (1.1 if c==ord('S') else 1/1.1)
        elif c in (ord('d'), ord('D')): chans[2].gain *= (1.1 if c==ord('D') else 1/1.1)
        elif c in (ord('f'), ord('F')): chans[3].gain *= (1.1 if c==ord('F') else 1/1.1)
        elif c in (ord('z'), ord('Z')): chans[0].offset += (0.05 if c==ord('Z') else -0.05)
        elif c in (ord('x'), ord('X')): chans[1].offset += (0.05 if c==ord('X') else -0.05)
        elif c in (ord('c'), ord('C')): chans[2].offset += (0.05 if c==ord('C') else -0.05)
        elif c in (ord('v'), ord('V')): chans[3].offset += (0.05 if c==ord('V') else -0.05)
        if not run: time.sleep(0.02)

# ---------- main ----------
def main():
    signal.signal(signal.SIGINT, lambda *_: sys.exit(0))
    curses.wrapper(draw_scope)

if __name__ == "__main__":
    main()

File: tscale/archive/old_implementations/ascii_scope.lua
-- Fixes for macOS/Lua 5.4 + robust nonblocking FIFO read + TTY-safe key input

local posix = require("posix")

-- ---------- config ----------
local PATH           = arg[1] or "/tmp/scope.fifo"
local MAX_CH         = 4
local REFRESH_HZ     = 30
local HIST_SEC       = 2.0
local DEFAULT_SPAN   = 0.250
local STACK_FRAC     = 0.45
local TRIG_POS_FRAC  = 0.25
local DIGITAL_LAST   = false

-- ---------- state ----------
local chans = {}
for i=1,MAX_CH do
  chans[i] = {visible=true, gain=1.0, offset=0.0}
end
local buf = {}
local buf_head = 0
local last_t = nil

local run = true
local time_span = DEFAULT_SPAN
local auto_span = true
local trig_enabled = false
local trig_ch = 1
local trig_lvl = 0.0
local trig_edge = 1
local trig_hyst = 1e-9
local use_envelope = true

local fd, ttyfd
local is_stdin = (PATH == "-" or PATH == "/dev/stdin")
local partial = ""   -- carry for split lines

-- ---------- term ----------
local function term_raw()
  -- macOS: stty -f /dev/tty; Linux: stty < /dev/tty
  if posix.uname().sysname == "Darwin" then
    os.execute("stty -f /dev/tty -icanon -echo min 0 time 0")
  else
    os.execute("stty -icanon -echo min 0 time 0 < /dev/tty")
  end
end
local function term_restore()
  if posix.uname().sysname == "Darwin" then
    os.execute("stty -f /dev/tty sane")
  else
    os.execute("stty sane < /dev/tty")
  end
end
local function cls() io.write("\27[2J") end
local function cup(r,c) io.write(string.format("\27[%d;%dH", r, c)) end
local function hide() io.write("\27[?25l") end
local function show() io.write("\27[?25h") end
local function rev(on) if on then io.write("\27[7m") else io.write("\27[27m") end end

-- ---------- util ----------
local function now()
  local ts = posix.clock_gettime("monotonic")
  return ts.tv_sec + ts.tv_nsec*1e-9
end
local function sleep(s) posix.nanosleep(s) end
local function split_ws(s)
  local t = {}
  for tok in string.gmatch(s, "%S+") do t[#t+1]=tok end
  return t
end
local function buf_push(t, vs) buf_head=buf_head+1; buf[buf_head]={t,vs} end
local function buf_trim(cut_t)
  local i=1
  while i<=buf_head and buf[i][1] < cut_t do buf[i]=nil; i=i+1 end
  if i>1 then
    local j=1
    for k=i,buf_head do buf[j]=buf[k]; buf[k]=nil; j=j+1 end
    buf_head=j-1
  end
end

-- ---------- IO ----------
local function open_tty()
  -- open /dev/tty for key input regardless of stdin source
  ttyfd = posix.open("/dev/tty", posix.O_RDONLY)
  if ttyfd then
    local fl = posix.fcntl(ttyfd, posix.F_GETFL)
    posix.fcntl(ttyfd, posix.F_SETFL, fl | posix.O_NONBLOCK) -- Lua 5.3+ bitwise |
  end
end

local function open_input()
  if is_stdin then
    fd = posix.fileno(io.stdin)
    local flags = posix.fcntl(fd, posix.F_GETFL)
    posix.fcntl(fd, posix.F_SETFL, flags | posix.O_NONBLOCK)
    return
  end
  local st = posix.stat(PATH)
  if not st then
    io.stderr:write("FIFO not found: ", PATH, " (mkfifo ", PATH, ")\n"); os.exit(1)
  end
  fd = posix.open(PATH, posix.O_RDONLY | posix.O_NONBLOCK)
end

local function parse_line(line)
  local parts = split_ws(line)
  if #parts < 2 then return nil end
  local t = tonumber(parts[1]); if not t then return nil end
  local vs = {}
  for i=2, math.min(1+MAX_CH, #parts) do
    vs[#vs+1] = tonumber(parts[i]) or 0.0
  end
  return t, vs
end

local function feed_data_chunk(chunk)
  partial = partial .. chunk
  local start = 1
  while true do
    local nl = string.find(partial, "\n", start, true)
    if not nl then
      -- retain remainder starting at 'start'
      partial = string.sub(partial, start)
      return
    end
    local line = string.sub(partial, start, nl-1)
    start = nl + 1
    if #line > 0 then
      local t, vs = parse_line(line)
      if t then
        if last_t and t < last_t then t = last_t + 1e-12 end
        last_t = t
        buf_push(t, vs)
      end
    end
  end
end

local function read_into_buffer()
  if not fd then return 0 end
  local rset = {[fd]=true}
  local rdy = {posix.select(rset, nil, nil, 0)}
  -- luaposix returns three tables; first is rset (possibly mutated)
  local rr = rdy[1]
  local n = 0
  if rr and rr[fd] then
    local data = posix.read(fd, 65536) or ""
    if #data == 0 then
      if not is_stdin then posix.close(fd); sleep(0.01); open_input() end
    else
      feed_data_chunk(data)
    end
  end
  if last_t then
    local cutoff = last_t - math.max(HIST_SEC, 5*DEFAULT_SPAN)
    buf_trim(cutoff)
    n = 1
  end
  return n
end

-- ---------- window ----------
local function estimate_dt()
  if buf_head < 5 then return nil end
  local a = math.max(1, buf_head - 200)
  local prev, n = nil, 0
  local dts = {}
  for i=a,buf_head do
    local t = buf[i][1]
    if prev and t>prev then dts[#dts+1] = t - prev end
    prev = t
  end
  if #dts == 0 then return nil end
  table.sort(dts)
  return dts[math.floor(#dts/2)+1]
end

local function compute_window(cols)
  if buf_head == 0 then
    local t = now()
    return t - time_span, t
  end
  if auto_span and buf_head > 2 then
    local dt = estimate_dt()
    if dt then
      local target_cols = math.max(10, math.floor(0.8*math.max(2, cols)))
      local span = math.max(5*dt, target_cols*dt)
      time_span = math.max(0.005, math.min(60.0, span))
    end
  end
  local right = buf[buf_head][1]
  return right - time_span, right
end

-- ---------- mapping/render ----------
local function map_row(val, rows, ci)
  local sep = rows / (MAX_CH + 1)
  local mid = math.floor(ci*sep + sep)
  local scale = sep * STACK_FRAC
  local y = math.floor(mid - val*scale + 0.5)
  if y < 2 then y = 2 end
  if y > rows then y = rows end
  return y
end

local function header_line(cols)
  local run_s = run and "ON" or "OFF"
  local edge_s = (trig_edge>0 and "+") or (trig_edge<0 and "-") or "+/-"
  return string.format("[q] quit  [space] run=%s  [</>] span=%.3fs  [m] auto=%s  [t] trig=%s ch=%d lvl=%+.3g edge=%s  [o] env/pts",
    run_s, time_span, (auto_span and "ON" or "OFF"), (trig_enabled and "ON" or "OFF"), trig_ch, trig_lvl, edge_s)
end

local function draw_frame(rows, cols, left_t, right_t)
  cup(1,1); rev(true)
  local hl = header_line(cols)
  io.write(hl); if #hl < cols then io.write(string.rep(" ", cols-#hl)) end
  rev(false)
  for y=2,rows do cup(y,1); io.write("|") end
  cup(rows+1,1); io.write("+" .. string.rep("-", cols-2) .. ">")
  local span = math.max(1e-12, right_t - left_t)

  if use_envelope then
    for ci=1,MAX_CH do
      if chans[ci].visible then
        local ymin, ymax = {}, {}
        for i=1,buf_head do
          local t,vs = buf[i][1], buf[i][2]
          if t >= left_t and t <= right_t and #vs >= ci then
            local xf = (cols-3) * (t - left_t) / span
            local x = 2 + math.floor(xf + 0.5)
            if x >= 2 and x <= cols-1 then
              local yv = chans[ci].gain*vs[ci] + chans[ci].offset
              if DIGITAL_LAST and ci==MAX_CH then yv = (yv>=0.5) and 0.8 or -0.8 end
              local y = map_row(yv, rows-1, ci)
              ymin[x] = (ymin[x] and math.min(ymin[x], y)) or y
              ymax[x] = (ymax[x] and math.max(ymax[x], y)) or y
            end
          end
        end
        for x=2,cols-1 do
          local y0, y1 = ymin[x], ymax[x]
          if y0 then for y=y0,y1 do cup(y,x); io.write("|") end end
        end
      end
    end
  else
    for ci=1,MAX_CH do
      if chans[ci].visible then
        local lastx, lasty = nil, nil
        for i=1,buf_head do
          local t,vs = buf[i][1], buf[i][2]
          if t >= left_t and t <= right_t and #vs >= ci then
            local xf = 1 + (cols-2) * (t - left_t) / span
            local x = math.floor(xf + 0.5)
            if x >= 2 and x <= cols-1 then
              local yv = chans[ci].gain*vs[ci] + chans[ci].offset
              if DIGITAL_LAST and ci==MAX_CH then yv = (yv>=0.5) and 0.8 or -0.8 end
              local y = map_row(yv, rows-1, ci)
              cup(y,x); io.write("*")
              if lastx and x>lastx then
                local dx, dy = x-lastx, y-lasty
                for k=1,dx-1 do
                  local xi = lastx + k
                  local yi = lasty + math.floor(dy*k/dx + 0.5)
                  if xi>=2 and xi<=cols-1 and yi>=2 and yi<=rows-1 then cup(yi,xi); io.write(".") end
                end
              end
              lastx, lasty = x, y
            end
          end
        end
      end
    end
  end

  cup(rows,3)
  local meta = {}
  for i=1,MAX_CH do
    local ch = chans[i]
    meta[#meta+1] = string.format("ch%d:%s g=%.3g off=%+.3g", i, ch.visible and "on" or "off", ch.gain, ch.offset)
  end
  io.write(table.concat(meta, "  "))
end

-- ---------- keys ----------
local function read_key()
  if not ttyfd then return nil end
  local rset = {[ttyfd]=true}
  local rdy = {posix.select(rset,nil,nil,0)}
  local rr = rdy[1]
  if rr and rr[ttyfd] then
    local c = posix.read(ttyfd, 1)
    if c and #c>0 then return c end
  end
  return nil
end

local function handle_key(c)
  if not c then return end
  local b = string.byte(c)
  if c == "q" or b == 27 then return "quit" end
  if c == " " then run = not run end
  if c == ">" or c == "." then time_span = math.min(60.0, time_span*1.25) end
  if c == "<" or c == "," then time_span = math.max(0.005, time_span/1.25) end
  if c == "m" then auto_span = not auto_span end
  if c == "t" then trig_enabled = not trig_enabled end
  if c == "g" then trig_ch = (trig_ch % MAX_CH) + 1 end
  if c == "+" or c == "=" then trig_lvl = trig_lvl + 0.05 end
  if c == "-" then trig_lvl = trig_lvl - 0.05 end
  if c == "r" then trig_edge = 1 end
  if c == "f" then trig_edge = -1 end
  if c == "e" then trig_edge = 0 end
  if c == "o" then use_envelope = not use_envelope end
  if c >= "1" and c <= "4" then
    local i = string.byte(c) - string.byte("0")
    chans[i].visible = not chans[i].visible
  end
  if c == "a" or c == "A" then chans[1].gain = chans[1].gain * (c=="A" and 1.1 or 1/1.1) end
  if c == "s" or c == "S" then chans[2].gain = chans[2].gain * (c=="S" and 1.1 or 1/1.1) end
  if c == "d" or c == "D" then chans[3].gain = chans[3].gain * (c=="D" and 1.1 or 1/1.1) end
  if c == "f" or c == "F" then chans[4].gain = chans[4].gain * (c=="F" and 1.1 or 1/1.1) end
  if c == "z" or c == "Z" then chans[1].offset = chans[1].offset + (c=="Z" and 0.05 or -0.05) end
  if c == "x" or c == "X" then chans[2].offset = chans[2].offset + (c=="X" and 0.05 or -0.05) end
  if c == "c" or c == "C" then chans[3].offset = chans[3].offset + (c=="C" and 0.05 or -0.05) end
  if c == "v" or c == "V" then chans[4].offset = chans[4].offset + (c=="V" and 0.05 or -0.05) end
end

-- ---------- main loop ----------
local function main()
  open_input()
  open_tty()
  io.stdout:setvbuf("no")
  term_raw(); hide()

  local rows, cols = 0, 0
  local function get_size()
    local ok, ws = pcall(posix.tcgetwinsize, 1)
    if ok and ws and ws.ws_row and ws.ws_col then
      rows, cols = ws.ws_row-2, ws.ws_col
    else
      rows = (tonumber(os.getenv("LINES")) or 24) - 2
      cols = tonumber(os.getenv("COLUMNS")) or 80
    end
    if rows < 10 then rows = 10 end
    if cols < 40 then cols = 80 end
  end

  local frame_dt = 1.0/REFRESH_HZ
  while true do
    get_size()
    if run then read_into_buffer() end
    cls()
    local left_t, right_t = compute_window(cols)
    if right_t <= left_t then right_t = left_t + 1e-3 end
    draw_frame(rows, cols, left_t, right_t)
    local k = read_key()
    if handle_key(k) == "quit" then break end
    sleep(frame_dt)
  end
  show(); term_restore(); cup(rows+3,1)
end

posix.signal(posix.SIGINT, function() show(); term_restore(); os.exit(0) end)
main()
